File: src/C3.Infrastructure/Updates/UpdateCheckService.vb
Namespace Updates

    Public Enum UpdateCheckOutcome
        Failed = 0
        NoPublishedRelease
        UpToDate
        UpdateAvailable
    End Enum

    Public NotInheritable Class UpdateCheckResult

        Private Sub New(
                outcome As UpdateCheckOutcome,
                manifest As UpdateReleaseManifest,
                message As String,
                failureException As Exception)

            Me.Outcome = outcome
            Me.Manifest = manifest
            Me.Message = If(message, String.Empty)
            Me.FailureException = failureException
        End Sub

        Public ReadOnly Property Outcome As UpdateCheckOutcome

        Public ReadOnly Property Manifest As UpdateReleaseManifest

        Public ReadOnly Property Message As String

        Public ReadOnly Property FailureException As Exception

        Public ReadOnly Property IsSuccess As Boolean
            Get
                Return Outcome <> UpdateCheckOutcome.Failed
            End Get
        End Property

        Public ReadOnly Property IsUpdateAvailable As Boolean
            Get
                Return Outcome = UpdateCheckOutcome.UpdateAvailable
            End Get
        End Property

        Friend Shared Function Completed(
                outcome As UpdateCheckOutcome,
                manifest As UpdateReleaseManifest,
                message As String) As UpdateCheckResult

            Return New UpdateCheckResult(outcome, manifest, message, Nothing)
        End Function

        Friend Shared Function Failed(
                message As String,
                Optional failureException As Exception = Nothing) As UpdateCheckResult

            Return New UpdateCheckResult(
                UpdateCheckOutcome.Failed,
                Nothing,
                message,
                failureException)
        End Function

    End Class

    ''' <summary>
    ''' Coordinates manifest retrieval and applies the publication and SemVer
    ''' precedence rules shared by both Windows build lanes.
    ''' </summary>
    Public NotInheritable Class UpdateCheckService

        Private ReadOnly _source As IUpdateManifestSource

        Public Sub New()
            Me.New(New HttpUpdateManifestSource())
        End Sub

        Public Sub New(source As IUpdateManifestSource)
            If source Is Nothing Then
                Throw New ArgumentNullException("source")
            End If
            _source = source
        End Sub

        Public Function Check(
                feedUrl As String,
                currentInformationalVersion As String,
                expectedChannel As String) As UpdateCheckResult

            Dim currentIdentity As SemanticVersion = Nothing
            If Not SemanticVersion.TryParse(currentInformationalVersion, currentIdentity) Then
                Return UpdateCheckResult.Failed(
                    "The current C3 release identity is invalid.")
            End If

            Dim feedUri As Uri = Nothing
            If Not Uri.TryCreate(feedUrl, UriKind.Absolute, feedUri) Then
                Return UpdateCheckResult.Failed(
                    "The configured update manifest endpoint is invalid.")
            End If

            Dim endpointFailure As String = Nothing
            If Not UpdateEndpointPolicy.TryValidate(
                    feedUri,
                    expectedChannel,
                    endpointFailure) Then
                Return UpdateCheckResult.Failed(endpointFailure)
            End If

            Try
                Dim readResult As UpdateManifestReadResult = _source.Read(
                    feedUri,
                    expectedChannel)
                If Not readResult.IsSuccess Then
                    Return UpdateCheckResult.Failed(
                        readResult.FailureMessage,
                        readResult.FailureException)
                End If
                Return Evaluate(currentIdentity, readResult.Manifest)
            Catch ex As Exception
                Return UpdateCheckResult.Failed(
                    "The update manifest could not be downloaded.",
                    ex)
            End Try
        End Function

        Public Shared Function Evaluate(
                currentInformationalVersion As String,
                manifest As UpdateReleaseManifest) As UpdateCheckResult

            Dim currentIdentity As SemanticVersion = Nothing
            If Not SemanticVersion.TryParse(currentInformationalVersion, currentIdentity) Then
                Return UpdateCheckResult.Failed(
                    "The current C3 release identity is invalid.")
            End If
            Return Evaluate(currentIdentity, manifest)
        End Function

        Private Shared Function Evaluate(
                currentIdentity As SemanticVersion,
                manifest As UpdateReleaseManifest) As UpdateCheckResult

            If manifest Is Nothing Then
                Return UpdateCheckResult.Failed("The update manifest is missing.")
            End If
            If Not manifest.Published Then
                Return UpdateCheckResult.Completed(
                    UpdateCheckOutcome.NoPublishedRelease,
                    manifest,
                    "The update channel does not currently advertise a published release.")
            End If
            If manifest.ReleaseIdentity.CompareTo(currentIdentity) > 0 Then
                Return UpdateCheckResult.Completed(
                    UpdateCheckOutcome.UpdateAvailable,
                    manifest,
                    "A newer published release is available.")
            End If
            Return UpdateCheckResult.Completed(
                UpdateCheckOutcome.UpToDate,
                manifest,
                "No newer published release is available.")
        End Function

    End Class

End Namespace

File: src/C3.Infrastructure/Updates/UpdateCheckService.cs
using Microsoft.VisualBasic.CompilerServices;
using System;

namespace C3.Infrastructure.Updates
{
	public sealed class UpdateCheckService
	{
		private readonly IUpdateManifestSource _source;

		public UpdateCheckService()
			: this(new HttpUpdateManifestSource())
		{
		}

		public UpdateCheckService(IUpdateManifestSource source)
		{
			if (source == null)
			{
				throw new ArgumentNullException("source");
			}
			_source = source;
		}

		public UpdateCheckResult Check(string feedUrl, string currentInformationalVersion, string expectedChannel)
		{
			SemanticVersion currentIdentity = null;
			if (!SemanticVersion.TryParse(currentInformationalVersion, ref currentIdentity))
			{
				return UpdateCheckResult.Failed("The current C3 release identity is invalid.", null);
			}
			Uri feedUri = null;
			if (!Uri.TryCreate(feedUrl, UriKind.Absolute, out feedUri))
			{
				return UpdateCheckResult.Failed("The configured update manifest endpoint is invalid.", null);
			}
			string message = null;
			if (!UpdateEndpointPolicy.TryValidate(feedUri, expectedChannel, ref message))
			{
				return UpdateCheckResult.Failed(message, null);
			}
			try
			{
				UpdateManifestReadResult updateManifestReadResult = _source.Read(feedUri, expectedChannel);
				if (!updateManifestReadResult.IsSuccess)
				{
					return UpdateCheckResult.Failed(updateManifestReadResult.FailureMessage, updateManifestReadResult.FailureException);
				}
				return Evaluate(currentIdentity, updateManifestReadResult.Manifest);
			}
			catch (Exception ex)
			{
				ProjectData.SetProjectError(ex);
				Exception failureException = ex;
				UpdateCheckResult result = UpdateCheckResult.Failed("The update manifest could not be downloaded.", failureException);
				ProjectData.ClearProjectError();
				return result;
			}
		}

		public static UpdateCheckResult Evaluate(string currentInformationalVersion, UpdateReleaseManifest manifest)
		{
			SemanticVersion currentIdentity = null;
			if (!SemanticVersion.TryParse(currentInformationalVersion, ref currentIdentity))
			{
				return UpdateCheckResult.Failed("The current C3 release identity is invalid.", null);
			}
			return Evaluate(currentIdentity, manifest);
		}

		private static UpdateCheckResult Evaluate(SemanticVersion currentIdentity, UpdateReleaseManifest manifest)
		{
			if (manifest == null)
			{
				return UpdateCheckResult.Failed("The update manifest is missing.", null);
			}
			if (!manifest.Published)
			{
				return UpdateCheckResult.Completed(UpdateCheckOutcome.NoPublishedRelease, manifest, "The update channel does not currently advertise a published release.");
			}
			if (manifest.ReleaseIdentity.CompareTo(currentIdentity) > 0)
			{
				return UpdateCheckResult.Completed(UpdateCheckOutcome.UpdateAvailable, manifest, "A newer published release is available.");
			}
			return UpdateCheckResult.Completed(UpdateCheckOutcome.UpToDate, manifest, "No newer published release is available.");
		}
	}
}
